aspnet/fundamentals/middleware.rst

```diff
@@ -1,8 +1,176 @@
+
+.. literalinclude:: middleware/sample/src/MiddlewareSample/Startup.cs
+	:language: c#
+	:linenos:
+	:lines: 81-93
+	:emphasize-lines: 11
+	:dedent: 8
+
+In addition to path-based mapping, the ``MapWhen`` method supports predicate-based middleware branching, allowing separate pipelines to be constructed in a very flexible fashion. Any predicate of type ``Func<HttpContext, bool>`` can be used to map requests to a new branch of the pipeline. In the following example, a simple predicate is used to detect the presence of a querystring variable ``branch``:
+
+.. literalinclude:: middleware/sample/src/MiddlewareSample/Startup.cs
+	:language: c#
+	:linenos:
+	:lines: 95-113
+	:emphasize-lines: 5,11-13
+	:dedent: 8
+
+Using the configuration shown above, any request that includes a querystring value for ``branch`` will use the pipeline defined in the ``HandleBranch`` method (in this case, a response of "Branch used."). All other requests (that do not define a querystring value for ``branch``) will be handled by the delegate defined on line 17.
+
+Built-in middleware
+-------------------
+
+ASP.NET ships with the following middleware components:
+
+
+.. list-table:: Middleware
+	:header-rows: 1
+	
+	*  - Middleware
+	   - Description
+	*  - :doc:`Authentication </security/sociallogins>`
+	   - Provides authentication support.
+	*  - :doc:`diagnostics`
+	   - Includes support for error pages and runtime information.
+	*  - :doc:`routing`
+	   - Define and constrain request routes.
+	*  - :doc:`Static Files <static-files>`
+	   - Provides support for serving static files, and directory browsing.
+
+.. Do we need to list any of these? MVC, AuthorizeBasicMiddleware, Session, CORS, ResponseCaching
+
+
+Writing middleware
+------------------
+
+For more complex request handling functionality, the ASP.NET team recommends implementing the middleware in its own class, and exposing an ``IApplicationBuilder`` extension method that can be called from the ``Configure`` method. The simple logging middleware shown in the previous example can be converted into a middleware class that takes in the next ``RequestDelegate`` in its constructor and supports an ``Invoke`` method as shown:
+
+.. literalinclude:: middleware/sample/src/MiddlewareSample/RequestLoggerMiddleware.cs
+	:language: c#
+	:caption: RequestLoggerMiddleware.cs
+	:linenos:
+	:emphasize-lines: 13, 19
+
+The middleware follows the `Explicit Dependencies Principle <http://deviq.com/explicit-dependencies-principle/>`_ and exposes all of its dependencies in its constructor. The extension method is responsible for providing the required dependencies. You can create overloads of the extension methods that allow dependencies to be passed in from the ``Configure`` method, as well. The ``UseRequestLogger`` extension method is shown below:
```

Comment:
Middleware can take advantage of the `UseMiddleware<T>` extension to inject DI services directly into their constructors. See:
https://github.com/aspnet/StaticFiles/blob/1.0.0-beta5/src/Microsoft.AspNet.StaticFiles/StaticFileExtensions.cs#L44
https://github.com/aspnet/StaticFiles/blob/1.0.0-beta5/src/Microsoft.AspNet.StaticFiles/StaticFileMiddleware.cs#L30
Any non-DI parameters can also be passed in so long as they come after the DI parameters in the constructor.